### PR TITLE
Factor out Travis 'install' sections for use with GitHub Actions

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

We need a common code path on GitHub Actions and Travis for Windows builds.

This PR achieves that for the `install:` sections on Travis.

## Related issue number

#631

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: java/runtime/src/main/java/io/ray/runtime/runner/RunManager.java [hunk 1]
```diff
@@ -84,7 +84,7 @@ public void terminateProcess(String name, Process p) {
       }
       numAttempts++;
     }
-    LOGGER.info("Process {} is now terminated.", name);
+    LOGGER.debug("Process {} is now terminated.", name);
```
Comment (raulchen) [inline java/runtime/src/main/java/io/ray/runtime/runner/RunManager.java:87]: I think these logs are useful and need to be info. Why change them to debug?

Comment (mehrdadn) [reply java/runtime/src/main/java/io/ray/runtime/runner/RunManager.java:87]: I feel the complete opposite to be honest. They've given me a really hard time seeing the actual errors buried in the middle (in fact for a long time I didn't even realize there might be errors in the middle). The non-Java jobs aren't nearly this verbose... compare [this](https://travis-ci.com/github/ray-project/ray/jobs/318669233) with [this](https://travis-ci.com/github/ray-project/ray/jobs/318669229) for instance.

(I wouldn't be surprised if this is one reason the Java builds have broken in the past by the way—the easier it is to overlook errors and the harder it is to find them, the more likely it is the build will break...)

File: ci/travis/ci.sh [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 set -eo pipefail && if [ -n "${OSTYPE##darwin*}" ]; then set -ux; fi  # some options interfere with Travis's RVM on Mac
 
-ROOT_DIR="$(cd "$(dirname "${BASH_SOURCE:-$0}")"; pwd)"
+ROOT_DIR="$(builtin cd "$(dirname "${BASH_SOURCE:-$0}")"; pwd)"
```
Comment (simon-mo) [inline ci/travis/ci.sh:7]: This is not maintainable. Developers can't remember to always use `builtin cd` for this when writing CI script. Is there another way to do this? What's the RVM issue?

File: java/runtime/src/main/java/io/ray/runtime/RayNativeRuntime.java [hunk 3]
```diff
@@ -96,7 +96,7 @@ public void start() {
     objectStore = new NativeObjectStore(workerContext);
     taskSubmitter = new NativeTaskSubmitter();
 
-    LOGGER.info("RayNativeRuntime started with store {}, raylet {}",
```
Comment (simon-mo) [inline java/runtime/src/main/java/io/ray/runtime/RayNativeRuntime.java:99]: can we revert this commit 0a743ff5889ae76875a86c0c27c17f1ee67e78d2? java logging is controlled separately in log4j config file. the write way is to change the log threshold in CI, not here.

File: ci/travis/ci.sh [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 { SHELLOPTS_STACK="${SHELLOPTS_STACK-}|$(set +o); set -$-"; } 2> /dev/null  # Push caller's shell options (quietly)
 
-set -exo pipefail
+set -eo pipefail && if [ -n "${OSTYPE##darwin*}" ]; then set -ux; fi  # some options interfere with Travis's RVM on Mac
```
Comment (simon-mo) [inline ci/travis/ci.sh:5]: just like `builtin cd`, this is not really maintenable. Is there a way to set these in one place? something like a new file
`prelude.sh` and then `source prelude.sh` in each script.

File: BUILD.bazel [hunk 5]
```diff
@@ -1341,7 +1341,11 @@ genrule(
             tmpdir="redis.tmp" &&
```
Comment (simon-mo) [inline BUILD.bazel:1341]: instead of this, maybe convert the whole thing to a shell script and then pipe the whole thing to dev null?

File: ci/travis/determine_tests_to_run.py [hunk 6]
```diff
@@ -149,25 +149,30 @@ def list_changed_files(commit_range):
         RAY_CI_ONLY_RLLIB_AFFECTED = 1
 
     # Log the modified environment variables visible in console.
-    for output_stream in [sys.stdout, sys.stderr]:
-        _print = partial(print, file=output_stream)
-        _print("export RAY_CI_TUNE_AFFECTED={}".format(RAY_CI_TUNE_AFFECTED))
-        _print("export RAY_CI_ONLY_RLLIB_AFFECTED={}"
+    for output_stream in [sys.stdout]:
```
Comment (simon-mo) [inline ci/travis/determine_tests_to_run.py:152]: `for output_stream in [sys.stdout, sys.stderr]:` is intentional to help debug which files are affected and the the env var. It's quite important.

File: ci/travis/install-bazel.sh [hunk 7]
```diff
@@ -85,6 +83,7 @@ if [ "${TRAVIS-}" = true ] || [ -n "${GITHUB_WORKFLOW-}" ]; then
 # CI output doesn't scroll, so don't use curses
 build --color=yes
 build --curses=no
+build --disk_cache="$(test "${OSTYPE}" = msys || echo ~/ray-bazel-cache)"
```
Comment (simon-mo) [inline ci/travis/install-bazel.sh:86]: 🧐what do we need the `test` condition?

File: ci/travis/install-dependencies.sh [hunk 8]
```diff
@@ -1,122 +1,169 @@
...
+  install_base
+  if [ -n "${GITHUB_WORKFLOW-}" ]; then  # Keep Travis's built-in compilers and only use this for GitHub Actions (for now)
+    . "${ROOT_DIR}"/install-toolchains.sh
+  fi
+  install_nvm
+  install_pip
+
+  if [ -n "${PYTHON-}" ]; then
+    install_miniconda
+    pip_packages=(scipy tensorflow=="${TF_VERSION-2.0.0b1}" cython==0.29.0 gym opencv-python-headless pyyaml \
```
Comment (simon-mo) [inline ci/travis/install-dependencies.sh:132]: is TF_VERSION-2.0.0b1 still set?

**Review by simon-mo — COMMENTED** [review]
Looks good overall. Can we write a quick `README.md` in `./ci` folder so devs knows:
- (High level) how each script interact with each other, how are they called? e.g. https://github.com/ucbrise/clipper/blob/develop/bin/README.md
- Where to add pip dependencies?
- Where to add system dependencies?

File: ci/travis/install-bazel.sh (hunk 7 above)
Comment (mehrdadn) [reply ci/travis/install-bazel.sh:86]: It's because I get random permission-denied errors on Windows that I haven't been able to figure out yet :\ I can look into it down the road but for now I'm just disabling the disk cache on Windows.

File: ci/travis/install-dependencies.sh (hunk 8 above)
Comment (mehrdadn) [reply ci/travis/install-dependencies.sh:132]: It's the same behavior as before, where it's (still) set only in the Travis YML file. The previous code here was just

    tf_version="$TF_VERSION"
    if [[ $tf_version == "" ]]; then tf_version="2.0.0b1"; fi

which this is equivalent to. Later on when migrating the tests to GitHub Actions we'll have to set the environment variables somehow, but this PR doesn't do that yet.

File: ci/travis/determine_tests_to_run.py (hunk 6 above)
Comment (mehrdadn) [reply ci/travis/determine_tests_to_run.py:152]: Yeah but it's [already being printed by the parent script](https://travis-ci.com/github/ray-project/ray/jobs/318792471#L213) when `set -x` is enabled; this part is redundant (floods the screen).

File: BUILD.bazel (hunk 5 above)
Comment (mehrdadn) [reply BUILD.bazel:1341]: I'm a little hesitant to do that in this PR for a few different reasons:

- My goal here was just to make the CI usable, not to refactor the Bazel `genrules`—that is something we should do down the road too, but it needs to happen for all them. Some are more complicated than this one, and in some cases there are also Bazel functions we should probably use (needs adding some dependencies) to e.g. copy files without having to write scripts.

- It would become potentially more brittle and less readable since `$(location ...)` would need to be passed as parameters.

- I only want to suppress progress messages and warnings (and only the ones that seem safe to suppress), so piping everything to `/dev/null` feels a little risky to me.

Comment (simon-mo) [reply BUILD.bazel:1341]: maybe we can just use `./suppress_output` script to suppress it for success run and print out the error on failure?

File: ci/travis/determine_tests_to_run.py (hunk 6 above)
Comment (simon-mo) [reply ci/travis/determine_tests_to_run.py:152]: oh ok i didn't notice that. in this case can you just change `_print` to normal print and remove the for loop?

File: ci/travis/install-bazel.sh (hunk 7 above)
Comment (simon-mo) [reply ci/travis/install-bazel.sh:86]: didn't we have bazel cache before for windows?

File: ci/travis/install-dependencies.sh (hunk 8 above)
Comment (simon-mo) [reply ci/travis/install-dependencies.sh:132]: but `TF_VERSION-2.0.0b1` is not set anywhere now?

File: ci/travis/ci.sh (hunk 2 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:7]: There are a couple issues to deal with:

- Travis on Mac at least seems to use RVM (a Ruby thing?) to hook `cd` for whatever reason, so whenever you run `cd` with `set -x`, it floods the screen with a ton of useless messages. `builtin cd` bypasses that.

- Some of those scripts also can't function with `set -u`; they error out and cause the scripts to fail, so we have to restore the `set -u` state before returning to the caller.

Forgetting `builtin` is already not a _big_ deal, since it merely floods the screen. The part about `set -u` is harder to get around though. I'll continue on that below. (...)

File: ci/travis/ci.sh (hunk 4 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:5]: Let's chat more about this, I agree it's not great but all the options have downsides. One trouble is that depends on the current directory, which could be different from the script directory. If I did `source "${ROOT_DIR}"/ci/prelude.sh` then that itself would require doing the `cd` business first to get `ROOT_DIR` which defeats the purpose.

It might be worth noting I _do_ hope to move the other scripts into `ci.sh` later, so this line would ideally only be present in this file. In fact I had already done that in this PR at first, but it made it difficult to review (and revert later potentially), so that's why I didn't do that here. But I think all the CI stuff should be in `ci.sh` for multiple reasons (factoring out this bit being being one of them, but not the only one).

File: BUILD.bazel (hunk 5 above)
Comment (mehrdadn) [reply BUILD.bazel:1341]: That still suppresses everything unless there's an actual error. Do you feel that's fine? It felt risky to me since I'm not sure what kinds of errors might be printed, but for example, seeing the linker warning about `tmpnam` being unsafe made me realize there's a flag to make it us `mkstemp` again, so it was actually beneficial I think.

File: ci/travis/determine_tests_to_run.py (hunk 6 above)
Comment (mehrdadn) [reply ci/travis/determine_tests_to_run.py:152]: Okay sure!

File: ci/travis/install-bazel.sh (hunk 7 above)
Comment (mehrdadn) [reply ci/travis/install-bazel.sh:86]: I've had `disk_cache` locally, but I don't think we ever had it on CI for Windows. The option previously was [just a few lines higher](https://github.com/ray-project/ray/pull/7988/files/a53b37ed9887ec1069562caa492a40a3420cd2ec..76c3150ca779ad98f9cf9425fa9564faf785f13d#diff-3b37d271915170b4a4a8a68fa2fb83f8L64) under `
if [ "${TRAVIS-}" = true ]; then ...`.

File: ci/travis/install-dependencies.sh (hunk 8 above)
Comment (mehrdadn) [reply ci/travis/install-dependencies.sh:132]: Oh, that's not the variable name :) the notation `${TF_VERSION-2.0.0b1}` means "`${TF_VERSION}` if `TF_VERSION` is set, otherwise `2.0.0b1`. `TF_VERSION` is (still) set in `.travis.yml` like before; I haven't touched it. So it's (almost) equivalent to what the old code was doing. (I could write `${TF_VERSION:-2.0.0b1}` to make it "is set and non-blank" to make it exactly equivalent but this seemed fine too.)

File: java/runtime/src/main/java/io/ray/runtime/RayNativeRuntime.java (hunk 3 above)
Comment (mehrdadn) [reply java/runtime/src/main/java/io/ray/runtime/RayNativeRuntime.java:99]: Hmm I'll message you, I'm not sure what the best solution is. Some of the `info` logging is useful but some of it is redundant, and changing it all might not give the best result.

File: ci/travis/ci.sh [hunk 9]
```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env bash
+
+{ SHELLOPTS_STACK="${SHELLOPTS_STACK-}|$(set +o); set -$-"; } 2> /dev/null  # Push caller's shell options (quietly)
```
Comment (simon-mo) [inline ci/travis/ci.sh:3]: move this to function and call it. 
also why is this necessary?

File: ci/travis/ci.sh [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+unset -f cd  # Travis defines this on Mac for RVM, but it floods the trace log and isn't relevant for us
+
+set -eo pipefail && if [ -n "${OSTYPE##darwin*}" ]; then set -ux; fi  # some options interfere with Travis's RVM on Mac
+
+ROOT_DIR="$(cd "$(dirname "${BASH_SOURCE:-$0}")"; pwd)"
+WORKSPACE_DIR="${ROOT_DIR}/../.."
+
+should_run_job() {
```
Comment (simon-mo) [inline ci/travis/ci.sh:12]: top level comment on what this function do?
let's do this https://google.github.io/styleguide/shellguide.html#function-comments

File: ci/travis/install-dependencies.sh [hunk 11]
```diff
@@ -1,122 +1,173 @@
...
+      ;;
+    *)
+      "${miniconda_target}" -b -p "${miniconda_dir}" | grep --line-buffered -v "^\(installing: \|installation finished\.\)"
+      ;;
+  esac
+  { local set_x="${-//[^x]/}"; } 2> /dev/null  # save set -x to suppress noise
+  set +x
+  local source_line='PYTHON3_BIN_PATH=python; PATH="$HOME/miniconda/bin:$PATH"; export PYTHON3_BIN_PATH PATH;'
+  test -f ~/.bashrc && grep -x -q -F "${source_line}" -- ~/.bashrc || echo "${source_line}" >> ~/.bashrc
+  test -z "${set_x}" || set -x  # restore set -x
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply ci/travis/install-dependencies.sh:59]: Sorry the wording is confusing, I meant that it restores the _state_ of `set -x`. I should've probably said "restore xtrace". So if it was disabled before the push then it'll stay disabled after the pop.

File: ci/travis/install-dependencies.sh [hunk 12]
```diff
@@ -1,122 +1,173 @@
...
+          unzip -q -- "${target}"
+          rm -f -- "${target}"
+          printf "%s\r\n" "root: $(cygpath -w -- "${NVM_HOME}")" "path: ${NVM_SYMLINK}" > settings.txt
+        )
+        printf "%s\n" "export NVM_HOME=\"$(cygpath -w -- "${NVM_HOME}")\"" 'nvm() { "${NVM_HOME}/nvm.exe" "$@"; }' > "${NVM_HOME}/nvm.sh"
+      fi
+  fi
+}
+
+install_pip() {
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply ci/travis/install-dependencies.sh:83]: Right but Conda isn't always installed (notice it's only installed under `PYTHON` or `LINT` right now), whereas we need `pip` in _all_ situations, so I can't just rely on Conda. (I think this was the existing behavior. But I think it's good because the installation is slow, too, so I'd rather not install it just for pip.)

File: ci/travis/install-dependencies.sh [hunk 13]
```diff
@@ -1,122 +1,173 @@
...
+      set +x # suppress set -x since it'll get very noisy here
+      . "${HOME}/.nvm/nvm.sh"
+      nvm install node
+      nvm use --silent node
+      npm config set loglevel warn  # make NPM quieter
+    )
+  fi
+}
+
+run_npm() {
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply ci/travis/install-dependencies.sh:117]: I can't, because `install-dependencies.sh` is sourced, but `install-ray.sh` isn't.

But I also tried not to move around code in this PR so that we don't have migrations and modifications happening at the same time. I can move around code in a later PR though.

File: .travis.yml [hunk 14]
```diff
@@ -8,25 +8,34 @@ matrix:
       env:
         - PYTHON=3.6
         - PYTHONWARNINGS=ignore
+        - RAY_DEFAULT_BUILD=1
+        - RAY_CYTHON_EXAMPLES=1
+      before_install:
+        - . ./ci/travis/ci.sh prepare RAY_CI_SERVE_AFFECTED,RAY_CI_TUNE_AFFECTED,RAY_CI_PYTHON_AFFECTED
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply .travis.yml:14]: I can if you really prefer that, but I was hoping trying not to add more environment variables, since they make the code really rigid and and have already been giving me a hard time.

File: ci/travis/determine_tests_to_run.py [hunk 15]
```diff
@@ -138,25 +148,20 @@ def list_changed_files(commit_range):
...
-               .format(RAY_CI_LINUX_WHEELS_AFFECTED))
-        _print("export RAY_CI_MACOS_WHEELS_AFFECTED={}"
-               .format(RAY_CI_MACOS_WHEELS_AFFECTED))
-        _print("export RAY_CI_STREAMING_CPP_AFFECTED={}"
-               .format(RAY_CI_STREAMING_CPP_AFFECTED))
-        _print("export RAY_CI_STREAMING_PYTHON_AFFECTED={}"
-               .format(RAY_CI_STREAMING_PYTHON_AFFECTED))
-        _print("export RAY_CI_STREAMING_JAVA_AFFECTED={}"
-               .format(RAY_CI_STREAMING_JAVA_AFFECTED))
+    print("{{ export {}; }} 2> /dev/null".format(" ".join([
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply ci/travis/determine_tests_to_run.py:151]: It suppresses a redundant print due to `set -x`. The `eval` will itself print the command so we don't need the `export` to do it again.

File: ci/travis/determine_tests_to_run.py [hunk 16]
```diff
@@ -25,7 +25,7 @@ def list_changed_files(commit_range):
         list: List of changed files within the commit range
     """
 
-    command = ["git", "diff", "--name-only", commit_range]
+    command = ["git", "diff", "--name-only", commit_range, "--"]
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply ci/travis/determine_tests_to_run.py:28]: Nothing major, it just makes the error messages a bit clearer since it's no longer ambiguous whether the argument is referring to commits or files.

File: ci/travis/ci.sh (hunk 9 above)
Comment (mehrdadn) [reply ci/travis/ci.sh:3]: It's necessary because the script is sourced, and modified the caller's state (`set -e`, `set -u`, etc.), and the caller (Travis in this case) isn't always compatible with this script's flags.

File: ci/travis/install-bazel.sh (hunk 7 above)
Comment (mehrdadn) [reply ci/travis/install-bazel.sh:84]: It's likely due to the remote cache. We've always had that.

File: BUILD.bazel (hunk 5 above)
Comment (mehrdadn) [reply BUILD.bazel:1341]: But if I print them when there is a warning then it would print exactly the same warnings I'm trying to suppress here though :(

I can try to simplify the sed calls, maybe that'll make them easier to understand?

File: ci/travis/install-dependencies.sh [hunk 17]
```diff
@@ -1,122 +1,173 @@
...
-
-if [[ "$PYTHON" == "3.6" ]] || [[ "$MAC_WHEELS" == "1" ]]; then
-  # Install the latest version of Node.js in order to build the dashboard.
-  source "$HOME/.nvm/nvm.sh"
-  nvm install node
-fi
-
-pip install psutil setproctitle \
-        --target="$ROOT_DIR/../../python/ray/thirdparty_files"
+{ SHELLOPTS_STACK="${SHELLOPTS_STACK-}|$(set +o); set -$-"; } 2> /dev/null  # Push caller's shell options (quietly)
```
[the remark this answers is not in the document]
Comment (mehrdadn) [reply ci/travis/install-dependencies.sh:3]: Push/pop are only relevant for scripts that have callers to return to (i.e., which are sourced rather than executed). Most scripts are just executed directly, but this script and `ci.sh` are sourced, hence they need to restore the state for Travis. I could turn these into functions, but there are 2 problems:

- The push/pop need to be the _first_ things that happens in the script (even surrounding the `ROOT_DIR=...` stuff) to avoid defeating their purpose.

- The push/pop in a callee will overwrite the one in a caller (unless I use different names for the same functionality, which I think might then make them confusing).

I was hoping to instead move `install-dependencies.sh` into `ci.sh` in a following PR, so that might make this easier to deal with? That way only that script will be sourced and need to have this line.

File: ci/travis/install-dependencies.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (mehrdadn) [reply ci/travis/install-dependencies.sh:117]: Wait sorry, I mixed this up with a different function. I can probably move this one :) thanks!

File: ci/travis/determine_tests_to_run.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (mehrdadn) [reply ci/travis/determine_tests_to_run.py:151]: Actually I just realized I should probably move that to the Bash side, let me do that.

File: BUILD.bazel [hunk 18]
```diff
@@ -1338,15 +1338,23 @@ genrule(
...
-            set -x &&
-            tmpdir="redis.tmp" &&
-            path=$(location @com_github_antirez_redis//:file) &&
-            cp -p -L -R -- "$${path%/*}" "$${tmpdir}" &&
-            make -s -C "$${tmpdir}" -j"$$(getconf _NPROCESSORS_ONLN || echo 1)" &&
-            mv "$${tmpdir}"/src/redis-server $(location redis-server) &&
-            chmod +x $(location redis-server) &&
-            mv "$${tmpdir}"/src/redis-cli $(location redis-cli) &&
-            chmod +x $(location redis-cli) &&
+            set -e
```
Comment (simon-mo) [inline BUILD.bazel:1341]: can you please revert this change? modifying makefiles and others are not good solution. It takes two pages of the screen max.

**Review by simon-mo — APPROVED** [review]

File: BUILD.bazel (hunk 18 above)
Comment (mehrdadn) [reply BUILD.bazel]: Okay sure, just reverted.
